Kernel/Devices/GPU/3dfx/VoodooDisplayConnector.cpp

```diff
@@ -0,0 +1,423 @@
+        | VIDPROCCFG_DesktopPixelFormat32Bit
+        | VIDPROCCFG_DesktopCLUTBypass;
+
+    // We only want to touch the 2X flag of the DAC Mode register, the other flags are preserved
+    regs.dac_mode = m_registers->dac_mode & ~DACMODE_2X;
+
+    u32 const max_pixel_clock_in_khz = 270000;
+    if (mode_setting.pixel_clock_in_khz > max_pixel_clock_in_khz) {
+        return ENOTSUP;
+    }
+
+    u32 horizontal_divisor = 8;
+    if (mode_setting.pixel_clock_in_khz > max_pixel_clock_in_khz / 2) {
+        horizontal_divisor = 16;
+        regs.dac_mode |= DACMODE_2X;
+        regs.vid_proc_cfg |= VIDPROCCFG_2XMode;
+    }
+
+    dbgln_if(TDFX_DEBUG, "3dfx: Calculating best PLL settings for pixel clock {} KHz", mode_setting.pixel_clock_in_khz);
+    auto pll = calculate_pll(mode_setting.pixel_clock_in_khz);
+    dbgln_if(TDFX_DEBUG, "3dfx: Best matching PLL settings: m={}, n={}, k={}. Frequency: {} KHz", pll.m, pll.n, pll.k, pll.frequency_in_khz());
+    regs.pll_ctrl0 = pll.register_value();
+    regs.vid_screen_size = width | (height << 12);
+    regs.vid_desktop_overlay_stride = mode_setting.horizontal_stride;
+    regs.misc_out_reg = MISCOUTPUT_ClockSelectPLL | MISCOUTPUT_CRTCAddressColor;
+    if (height < 768) {
+        regs.misc_out_reg |= MISCOUTPUT_VerticalSyncPositive | MISCOUTPUT_HorizontalSyncPositive;
+    }
+
+    u32 hor_total = mode_setting.horizontal_total() / horizontal_divisor - 5;
+    u32 hor_disp_en_end = width / horizontal_divisor - 1;
+    u32 hor_sync_start = mode_setting.horizontal_sync_start() / horizontal_divisor;
+    u32 hor_sync_end = (mode_setting.horizontal_sync_end() / horizontal_divisor) & 0x1f;
+    u32 hor_blank_start = hor_disp_en_end;
+    u32 hor_blank_end = hor_total & 0x7f;
+
+    u32 vert_total = mode_setting.vertical_total() - 2;
+    u32 vert_disp_en_end = height - 1;
+    u32 vert_sync_start = mode_setting.vertical_sync_start();
+    u32 vert_sync_end = mode_setting.vertical_sync_end() & 0xf;
+    u32 vert_blank_start = mode_setting.vertical_blanking_start() - 1;
+    u32 vert_blank_end = (mode_setting.vertical_total() - 1) & 0xff;
+
+    if (hor_total > 0x1ff ||        // 9-bit field
+        hor_disp_en_end > 0x1ff ||  // 9-bit field
+        hor_sync_start > 0x1ff ||   // 9-bit field
+        hor_sync_end > 0x1f ||      // 5-bit field
+        hor_blank_start > 0x1ff ||  // 9-bit field
+        hor_blank_end > 0x7f ||     // 7-bit field
+        vert_total > 0x7ff ||       // 11-bit field
+        vert_disp_en_end > 0x7ff || // 11-bit field
+        vert_sync_start > 0x7ff ||  // 11-bit field
+        vert_sync_end > 0x0f ||     // 4-bit field
+        vert_blank_start > 0x7ff || // 11-bit field
+        vert_blank_end > 0xff       // 8-bit field
+    ) {
+        dbgln_if(TDFX_DEBUG, "3dfx: One of the timing values is too large to fit in its register");
+        return EOVERFLOW;
+    }
+
+    // CRT Controller Registers
+    regs.cr[0x00] = hor_total;             // bit 0-7 of hor_total
```

Comment:
can we give these register fields names? an enum or a set of `constexpr` vaiables would do.

Reply:
I will give it a try and post below what it will look like.

I'm not sure it will provide clarity. The values are packed in such a way that the names often aren't directly very meaningful. It's a slightly extended version of the classic VGA registers, except accessed at a different location as specified in the PCI BAR.

For example, the name for `cr[0x5]` (aka CR5) is _Horizontal Sync End_ so you would expect it to container the value of `hor_sync_end`. It does, but only bits 0-4 of it, bit 5 is in CR1A and CR5 also contains bit 5 of `hor_blank_end` and another 2 bits for a field that is not being used here. Most of the `vert_` values are split into 4 parts. This is also the reason why I added all those comments describing what bits go where.

Reply:
It's better than I expected to be honest. I should probably change the names of things like `CR17_TimingEnable` to `CR_ModeControl_Timingenable` if it's like this. The tightly packed bits remain a mess but that's not a surprise. :sweat_smile: 

```cpp
    constexpr auto CR_HorizontalTotal = 0x00;
    constexpr auto CR_HorizontalDisplayEnableEnd = 0x01;
    constexpr auto CR_HorizontalBlankingStart = 0x02;
    constexpr auto CR_HorizontalBlankingEnd = 0x03;
    constexpr auto CR_HorizontalSyncStart = 0x04;
    constexpr auto CR_HorizontalSyncEnd = 0x05;
    constexpr auto CR_VerticalTotal = 0x06;
    constexpr auto CR_Overflow = 0x07;
    constexpr auto CR_MaximumScanLine = 0x09;
    constexpr auto CR_VerticalSyncStart = 0x10;
    constexpr auto CR_VerticalSyncEnd = 0x11;
    constexpr auto CR_VerticalDisplayEnableEnd = 0x12;
    constexpr auto CR_VerticalBlankingStart = 0x15;
    constexpr auto CR_VerticalBlankingEnd = 0x16;
    constexpr auto CR_ModeControl = 0x17;
    constexpr auto CR_HorizontalExtensions = 0x1a;
    constexpr auto CR_VerticalExtensions = 0x1b;

    constexpr auto GR_GraphicsControllerMiscellaneous = 0x06;

    constexpr auto AR_AttributeControllerMode = 0x10;

    constexpr auto SR_SequencerReset = 0x00;
    constexpr auto SR_SequencerClockingMode = 0x01;

    // CRT Controller Registers
    regs.cr[CR_HorizontalTotal] = hor_total;                   // bit 0-7 of hor_total
    regs.cr[CR_HorizontalDisplayEnableEnd] = hor_disp_en_end;  // bit 0-7 of hor_disp_en_end
    regs.cr[CR_HorizontalBlankingStart] = hor_blank_start;     // bit 0-7 of hor_blank_start
    regs.cr[CR_HorizontalBlankingEnd] = (hor_blank_end & 0x1f) // bit 0-4 of hor_blank_end
        | CR3_CompatibilityRead;
    regs.cr[CR_HorizontalSyncStart] = hor_sync_start;                // bit 0-7 of hor_sync_start
    regs.cr[CR_HorizontalSyncEnd] = ((hor_sync_end & 0x1f)           // bit 0-4 of hor_sync_end
        | ((hor_blank_end & 0x20) << 2));                            // bit 5 of hor_blank_end
    regs.cr[CR_VerticalTotal] = vert_total;                          // bit 0-7 of vert_total
    regs.cr[CR_Overflow] = (((vert_total & 0x100) >> 8)              // bit 8 of vert_total
        | ((vert_disp_en_end & 0x100) >> 7)                          // bit 8 of vert_disp_en_end
        | ((vert_sync_start & 0x100) >> 6)                           // bit 8 of vert_sync_start
        | ((vert_blank_start & 0x100) >> 5)                          // bit 8 of vert_blank_start
        | ((vert_total & 0x200) >> 4)                                // bit 9 of vert_disp_en_end
        | ((vert_disp_en_end & 0x200) >> 3)                          // bit 9 of vert_disp_en_end
        | ((vert_sync_start & 0x200) >> 2));                         // bit 9 of vert_sync_start
    regs.cr[CR_MaximumScanLine] = ((vert_blank_start & 0x200) >> 4); // bit 9 of vert_blank_start
    regs.cr[CR_VerticalSyncStart] = vert_sync_start;                 // bit 0-7 of vert_sync_start
    regs.cr[CR_VerticalSyncEnd] = (vert_sync_end & 0x0f)             // bit 0-3 of vert_sync_end
        | CR11_EnableVertInt;
    regs.cr[CR_VerticalDisplayEnableEnd] = vert_disp_en_end; // bit 0-7 of vert_disp_en_end
    regs.cr[CR_VerticalBlankingStart] = vert_blank_start;    // bit 0-7 of vert_blank_start
    regs.cr[CR_VerticalBlankingEnd] = vert_blank_end;        // bit 0-7 of vert_blank_end
    regs.cr[CR_ModeControl] = CR17_TimingEnable | CR17_ByteWordMode;
    regs.cr[CR_HorizontalExtensions] = (hor_total & 0x100) >> 8 // bit 8 of hor_total
        | (hor_disp_en_end & 0x100) >> 6                        // bit 8 of hor_disp_en_end
        | (hor_blank_start & 0x100) >> 4                        // bit 8 of hor_blank_start
        | (hor_blank_end & 0x40) >> 1                           // bit 6 of hor_blank_end
        | (hor_sync_start & 0x100) >> 2                         // bit 8 of hor_sync_start
        | (hor_sync_end & 0x20) << 2;                           // bit 5 of hor_sync_end
    regs.cr[CR_VerticalExtensions] = (vert_total & 0x400) >> 10 // bit 10 of vert_total
        | (vert_disp_en_end & 0x400) >> 8                       // bit 10 of vert_disp_en_endx
        | (vert_blank_start & 0x400) >> 6                       // bit 10 of vert_blank_start
        | (vert_blank_end & 0x400) >> 4                         // bit 10 of vert_blank_end
        | (vert_sync_start & 0x400) >> 4;                       // bit 10 of vert_sync_start
    // Graphics Controller Registers
    regs.gr[GR_GraphicsControllerMiscellaneous] = GR6_MemoryMapEGAVGAExtended;

    // Attribute Controller Registers
    regs.ar[AR_AttributeControllerMode] = AR10_GraphicsMode | AR10_PixelWidth;

    // Sequencer Registers
    regs.sr[SR_SequencerReset] = SR0_AsynchronousReset | SR0_SynchronousReset;
    regs.sr[SR_SequencerClockingMode] = SR1_DotClock8;

 ```

Reply:
That does look a lot better than the magic numbers. even better might be 

```
enum class CR {
   HorizontalTotal = 0,
   HorizontalDisplayEnableEnd = 0x01,
   ...
};
```

So that you refer to it as `cr[CR::HorizontalTotal]` and not do things with those names that are against our coding standard :)

Reply:
This really looks like a dedicated struct/bit-field would be nicer....

Reply:
Alright changed it to structs after all, and got rid of the underscores in the names where they should not be.

Reply:
The `enum class` would have required casts everywhere they're being used as an index. And for the flags they're not really suitable either because there's no `operator|` on them, so that would mean even more casts.

I don't like that I can't easily see the offsets now though.

Reply:
I added the offsets/mnemonics as comments